File: app/src/main/java/com/winlator/cmod/StoresFragment.java
package com.winlator.cmod;

import android.content.Intent;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;
import androidx.appcompat.widget.SwitchCompat;
import android.net.Uri;
import android.app.Activity;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.Fragment;

import com.winlator.cmod.steam.SteamLoginActivity;
import com.winlator.cmod.steam.service.SteamService;
import com.winlator.cmod.steam.utils.PrefManager;
import com.winlator.cmod.core.FileUtils;

/**
 * Fragment showing store sign-in / sign-out for Steam, Epic, GOG, Amazon.
 */
public class StoresFragment extends Fragment {
    
    private LinearLayout layout;
    private static final int REQUEST_CODE_DEFAULT_PATH = 2001;
    private static final int REQUEST_CODE_STEAM_PATH = 2002;
    private static final int REQUEST_CODE_EPIC_PATH = 2003;
    private static final int REQUEST_CODE_GOG_PATH = 2004;
    private static final int REQUEST_CODE_AMAZON_PATH = 2005;


    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        if (getActivity() != null && ((AppCompatActivity) getActivity()).getSupportActionBar() != null) {
            ((AppCompatActivity) getActivity()).getSupportActionBar().setTitle(R.string.stores);
        }
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        PrefManager.INSTANCE.init(getContext());
        this.layout = new LinearLayout(getContext());
        layout.setOrientation(LinearLayout.VERTICAL);
        int pad = dpToPx(20);
        layout.setPadding(pad, pad, pad, pad);

        android.widget.ScrollView scrollView = new android.widget.ScrollView(getContext());
        scrollView.addView(layout);
        return scrollView;
    }

    private void updateStores() {
        if (layout == null) return;
        layout.removeAllViews();

        addStoreRow(layout, "Steam",
                SteamService.Companion.isLoggedIn(),
                v -> {
                    if (SteamService.Companion.isLoggedIn()) {
                        SteamService.Companion.logOut();
                        updateStores();
                    } else {
                        startActivity(new Intent(getContext(), SteamLoginActivity.class));
                    }
                });

        boolean isEpicLoggedIn = com.winlator.cmod.epic.service.EpicAuthManager.isLoggedIn(getContext());
        addStoreRow(layout, "Epic Games", isEpicLoggedIn, v -> {
            if (isEpicLoggedIn) {
                com.winlator.cmod.epic.service.EpicAuthManager.logoutSync(getContext());
                updateStores();
            } else {
                startActivity(new Intent(getContext(), com.winlator.cmod.epic.ui.auth.EpicOAuthActivity.class));
            }
        });

        addStoreRow(layout, "GOG", false, v ->
                android.widget.Toast.makeText(getContext(), "GOG integration coming soon", android.widget.Toast.LENGTH_SHORT).show());

        addStoreRow(layout, "Amazon Games", false, v ->
                android.widget.Toast.makeText(getContext(), "Amazon Games integration coming soon", android.widget.Toast.LENGTH_SHORT).show());

        // Downloads section
        addHeader(layout, "DOWNLOADS");

        addToggleRow(layout, "Wi-Fi Only Downloads", PrefManager.INSTANCE.getDownloadOnWifiOnly(), (buttonView, isChecked) -> {
            PrefManager.INSTANCE.setDownloadOnWifiOnly(isChecked);
        });

        addToggleRow(layout, "Shared Downloads Folder", PrefManager.INSTANCE.getUseSingleDownloadFolder(), (buttonView, isChecked) -> {
            PrefManager.INSTANCE.setUseSingleDownloadFolder(isChecked);
            updateStores();
        });

        if (PrefManager.INSTANCE.getUseSingleDownloadFolder()) {
            addPathRow(layout, "Default Downloads Folder", PrefManager.INSTANCE.getDefaultDownloadFolder(), REQUEST_CODE_DEFAULT_PATH);
        } else {
            addPathRow(layout, "Steam Store Downloads Folder", PrefManager.INSTANCE.getSteamDownloadFolder(), REQUEST_CODE_STEAM_PATH);
            addPathRow(layout, "Epic Store Downloads Folder", PrefManager.INSTANCE.getEpicDownloadFolder(), REQUEST_CODE_EPIC_PATH);
            addPathRow(layout, "GOG Store Downloads Folder", PrefManager.INSTANCE.getGogDownloadFolder(), REQUEST_CODE_GOG_PATH);
            addPathRow(layout, "Amazon Store Downloads Folder", PrefManager.INSTANCE.getAmazonDownloadFolder(), REQUEST_CODE_AMAZON_PATH);
        }
    }

    private void addHeader(LinearLayout parent, String title) {
        TextView header = new TextView(getContext());
        header.setText(title);
        header.setTextSize(12);
        header.setTextColor(0xFF8B949E);
        header.setPadding(dpToPx(4), dpToPx(24), dpToPx(4), dpToPx(8));
        parent.addView(header);
    }

    private void addToggleRow(LinearLayout parent, String title, boolean isChecked, android.widget.CompoundButton.OnCheckedChangeListener listener) {
        LinearLayout row = new LinearLayout(getContext());
        row.setOrientation(LinearLayout.HORIZONTAL);
        row.setPadding(dpToPx(4), dpToPx(12), dpToPx(4), dpToPx(12));
        row.setGravity(android.view.Gravity.CENTER_VERTICAL);

        TextView nameView = new TextView(getContext());
        nameView.setText(title);
        nameView.setTextSize(14);
        nameView.setTextColor(0xFFE6EDF3);
        LinearLayout.LayoutParams nameParams = new LinearLayout.LayoutParams(0, LinearLayout.LayoutParams.WRAP_CONTENT, 1f);
        nameView.setLayoutParams(nameParams);
        row.addView(nameView);

        SwitchCompat switchView = new SwitchCompat(getContext());
        switchView.setChecked(isChecked);
        switchView.setOnCheckedChangeListener(listener);
        row.addView(switchView);

        parent.addView(row);

        View divider = new View(getContext());
        divider.setBackgroundColor(0xFF30363D);
        parent.addView(divider, new LinearLayout.LayoutParams(LinearLayout.LayoutParams.MATCH_PARENT, 1));
    }

    private void addPathRow(LinearLayout parent, String title, String currentPath, int requestCode) {
        LinearLayout container = new LinearLayout(getContext());
        container.setOrientation(LinearLayout.VERTICAL);
        container.setPadding(dpToPx(4), dpToPx(12), dpToPx(4), dpToPx(12));

        TextView titleView = new TextView(getContext());
        titleView.setText(title);
        titleView.setTextSize(14);
        titleView.setTextColor(0xFFE6EDF3);
        container.addView(titleView);

        LinearLayout row = new LinearLayout(getContext());
        row.setOrientation(LinearLayout.HORIZONTAL);
        row.setGravity(android.view.Gravity.CENTER_VERTICAL);
        row.setPadding(0, dpToPx(4), 0, 0);

        TextView pathView = new TextView(getContext());
        String displayPath = currentPath;
        if (displayPath == null || displayPath.isEmpty()) {
            displayPath = "Not set";
        } else {
            try {
                Uri uri = Uri.parse(displayPath);
                String decodedPath = FileUtils.getFilePathFromUri(getContext(), uri);
                if (decodedPath != null) displayPath = decodedPath;
            } catch (Exception e) {}
        }
        pathView.setText(displayPath);
        pathView.setTextSize(12);
        pathView.setTextColor(0xFF8B949E);
        LinearLayout.LayoutParams pathParams = new LinearLayout.LayoutParams(0, LinearLayout.LayoutParams.WRAP_CONTENT, 1f);
        pathView.setLayoutParams(pathParams);
        row.addView(pathView);

        TextView actionBtn = new TextView(getContext());
        actionBtn.setText("INSTALL"); // As requested: "By setting the default folder by pressing ( Install )"
        actionBtn.setTextSize(12);
        actionBtn.setTextColor(0xFF57CBDE);
        actionBtn.setPadding(dpToPx(12), dpToPx(6), dpToPx(12), dpToPx(6));
        actionBtn.setBackgroundColor(0x2057CBDE);
        actionBtn.setOnClickListener(v -> {
            Intent intent = new Intent(Intent.ACTION_OPEN_DOCUMENT_TREE);
            startActivityForResult(intent, requestCode);
        });
        row.addView(actionBtn);

        container.addView(row);
        parent.addView(container);

        View divider = new View(getContext());
        divider.setBackgroundColor(0xFF30363D);
        parent.addView(divider, new LinearLayout.LayoutParams(LinearLayout.LayoutParams.MATCH_PARENT, 1));
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (resultCode == Activity.RESULT_OK && data != null) {
            Uri uri = data.getData();
            if (uri != null) {
                try {
                    getContext().getContentResolver().takePersistableUriPermission(
                            uri,
                            Intent.FLAG_GRANT_READ_URI_PERMISSION | Intent.FLAG_GRANT_WRITE_URI_PERMISSION
                    );
                } catch (SecurityException e) {}

                String uriString = uri.toString();
                switch (requestCode) {
                    case REQUEST_CODE_DEFAULT_PATH: PrefManager.INSTANCE.setDefaultDownloadFolder(uriString); break;
                    case REQUEST_CODE_STEAM_PATH: PrefManager.INSTANCE.setSteamDownloadFolder(uriString); break;
                    case REQUEST_CODE_EPIC_PATH: PrefManager.INSTANCE.setEpicDownloadFolder(uriString); break;
                    case REQUEST_CODE_GOG_PATH: PrefManager.INSTANCE.setGogDownloadFolder(uriString); break;
                    case REQUEST_CODE_AMAZON_PATH: PrefManager.INSTANCE.setAmazonDownloadFolder(uriString); break;
                }
                updateStores();
            }
        }
    }

    @Override
    public void onResume() {
        super.onResume();
        updateStores();
    }

    private void refreshView() {
        if (getView() != null) {
            getParentFragmentManager().beginTransaction()
                    .detach(this)
                    .attach(this)
                    .commit();
        }
    }

    private void addStoreRow(LinearLayout parent, String storeName, boolean isLoggedIn, View.OnClickListener onClick) {
        LinearLayout row = new LinearLayout(getContext());
        row.setOrientation(LinearLayout.HORIZONTAL);
        row.setPadding(dpToPx(4), dpToPx(12), dpToPx(4), dpToPx(12));
        row.setGravity(android.view.Gravity.CENTER_VERTICAL);

        // Store name
        TextView nameView = new TextView(getContext());
        nameView.setText(storeName);
        nameView.setTextSize(15);
        nameView.setTextColor(0xFFE6EDF3);
        LinearLayout.LayoutParams nameParams = new LinearLayout.LayoutParams(0, LinearLayout.LayoutParams.WRAP_CONTENT, 1f);
        nameView.setLayoutParams(nameParams);
        row.addView(nameView);

        // Status dot + text
        TextView statusView = new TextView(getContext());
        statusView.setText(isLoggedIn ? "● Signed In" : "○ Not Signed In");
        statusView.setTextSize(12);
        statusView.setTextColor(isLoggedIn ? 0xFF57CBDE : 0xFF8B949E);
        statusView.setPadding(0, 0, dpToPx(12), 0);
        row.addView(statusView);

        // Action button
        TextView actionBtn = new TextView(getContext());
        actionBtn.setText(isLoggedIn ? "SIGN OUT" : "SIGN IN");
        actionBtn.setTextSize(12);
        actionBtn.setTextColor(isLoggedIn ? 0xFFFF6B6B : 0xFF57CBDE);
        actionBtn.setPadding(dpToPx(12), dpToPx(6), dpToPx(12), dpToPx(6));
        actionBtn.setBackgroundColor(isLoggedIn ? 0x20FF6B6B : 0x2057CBDE);
        actionBtn.setOnClickListener(onClick);
        row.addView(actionBtn);

        parent.addView(row);

        // Thin divider
        View divider = new View(getContext());
        divider.setBackgroundColor(0xFF30363D);
        parent.addView(divider, new LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.MATCH_PARENT, 1));
    }

    private int dpToPx(int dp) {
        return (int) (dp * getResources().getDisplayMetrics().density);
    }
}

File: app/src/main/java/com/winlator/cmod/StoresFragment.kt
/* Settings > Stores fragment — hosts StoresScreen via ComposeView. */
package com.winlator.cmod

import android.content.Intent
import android.graphics.drawable.GradientDrawable
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout
import android.widget.ScrollView
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AppCompatActivity
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.darkColorScheme
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.ComposeView
import androidx.compose.ui.platform.ViewCompositionStrategy
import androidx.fragment.app.Fragment
import com.winlator.cmod.core.FileUtils
import com.winlator.cmod.epic.service.EpicAuthManager
import com.winlator.cmod.epic.ui.auth.EpicOAuthActivity
import com.winlator.cmod.gog.service.GOGAuthManager
import com.winlator.cmod.gog.service.GOGService
import com.winlator.cmod.gog.ui.auth.GOGOAuthActivity
import com.winlator.cmod.steam.SteamLoginActivity
import com.winlator.cmod.steam.service.SteamService
import com.winlator.cmod.steam.utils.PrefManager
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import org.json.JSONObject

class StoresFragment : Fragment() {

    private var storeState by mutableStateOf(StoreState())
    private lateinit var serverOptions: List<Pair<Int, String>>

    // Folder-picker launchers
    private val defaultFolderLauncher = registerForActivityResult(
        ActivityResultContracts.OpenDocumentTree()
    ) { uri -> uri?.let { persistUri(it); PrefManager.defaultDownloadFolder = it.toString(); refresh() } }

    private val steamFolderLauncher = registerForActivityResult(
        ActivityResultContracts.OpenDocumentTree()
    ) { uri -> uri?.let { persistUri(it); PrefManager.steamDownloadFolder = it.toString(); refresh() } }

    private val epicFolderLauncher = registerForActivityResult(
        ActivityResultContracts.OpenDocumentTree()
    ) { uri -> uri?.let { persistUri(it); PrefManager.epicDownloadFolder = it.toString(); refresh() } }

    private val gogFolderLauncher = registerForActivityResult(
        ActivityResultContracts.OpenDocumentTree()
    ) { uri -> uri?.let { persistUri(it); PrefManager.gogDownloadFolder = it.toString(); refresh() } }

    private val amazonFolderLauncher = registerForActivityResult(
        ActivityResultContracts.OpenDocumentTree()
    ) { uri -> uri?.let { persistUri(it); PrefManager.amazonDownloadFolder = it.toString(); refresh() } }

    private val gogLoginLauncher = registerForActivityResult(
        ActivityResultContracts.StartActivityForResult()
    ) { result ->
        if (result.resultCode == android.app.Activity.RESULT_OK) {
            val code = result.data?.getStringExtra(GOGOAuthActivity.EXTRA_AUTH_CODE)
            if (!code.isNullOrBlank()) {
                CoroutineScope(Dispatchers.Main).launch {
                    val authResult = GOGAuthManager.authenticateWithCode(requireContext(), code)
                    if (authResult.isSuccess) {
                        GOGService.start(requireContext())
                    }
                    refresh()
                }
            }
        }
    }

    private val epicLoginLauncher = registerForActivityResult(
        ActivityResultContracts.StartActivityForResult()
    ) { result ->
        if (result.resultCode == android.app.Activity.RESULT_OK) {
            val code = result.data?.getStringExtra(EpicOAuthActivity.EXTRA_AUTH_CODE)
            if (!code.isNullOrBlank()) {
                CoroutineScope(Dispatchers.Main).launch {
                    EpicAuthManager.authenticateWithCode(requireContext(), code)
                    refresh()
                }
            } else {
                refresh()
            }
        } else {
            refresh()
        }
    }

    private val steamLoginLauncher = registerForActivityResult(
        ActivityResultContracts.StartActivityForResult()
    ) {
        refresh()
    }

    // Lifecycle
    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        (activity as? AppCompatActivity)?.supportActionBar?.setTitle(R.string.stores_accounts_title)
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?,
    ): View {
        PrefManager.init(requireContext())
        serverOptions = loadServerOptions()
        refresh()

        val composeView = ComposeView(requireContext()).apply {
            setViewCompositionStrategy(ViewCompositionStrategy.DisposeOnViewTreeLifecycleDestroyed)
            setContent {
                MaterialTheme(
                    colorScheme = darkColorScheme(
                        primary    = Color(0xFF1A9FFF),
                        background = Color(0xFF0D1117),
                        surface    = Color(0xFF161B22),
                    )
                ) {
                    StoresScreen(
                        state                = storeState,
                        serverOptions        = serverOptions,
                        onSteamSignIn        = { steamLoginLauncher.launch(Intent(requireContext(), SteamLoginActivity::class.java)) },
                        onSteamSignOut       = {
                            SteamService.logOut()
                            refresh()
                        },
                        onEpicSignIn         = { epicLoginLauncher.launch(Intent(requireContext(), EpicOAuthActivity::class.java)) },
                        onEpicSignOut        = { EpicAuthManager.logoutSync(requireContext()); refresh() },
                        onGogSignIn          = { gogLoginLauncher.launch(Intent(requireContext(), GOGOAuthActivity::class.java)) },
                        onGogSignOut         = {
                            CoroutineScope(Dispatchers.Main).launch {
                                GOGService.logout(requireContext())
                                refresh()
                            }
                        },
                        onSharedFolderChanged = { PrefManager.useSingleDownloadFolder = it; refresh() },
                        onDownloadSpeedChanged = { PrefManager.downloadSpeed = it; refresh() },
                        onDownloadServerChanged = { cellId ->
                            PrefManager.cellId = cellId
                            PrefManager.cellIdManuallySet = cellId != 0
                            refresh()
                        },
                        onPickDefaultFolder  = { defaultFolderLauncher.launch(null) },
                        onPickSteamFolder    = { steamFolderLauncher.launch(null) },
                        onPickEpicFolder     = { epicFolderLauncher.launch(null) },
                        onPickGogFolder      = { gogFolderLauncher.launch(null) },
                        onPickAmazonFolder   = { amazonFolderLauncher.launch(null) },
                    )
                }
            }
        }

        // View level ScrollView gives Compose a bounded height required because ComposeView
        // sits inside a weighted LinearLayout that passes unbounded constraints to its children.
        // WARNING: Do NOT add verticalScroll/fillMaxSize to StoresScreen while this is here;
        // fillMaxSize inside an unbounded container crashes at runtime.
        // TODO: Remove this block and use Compose native scrolling once fully migrated off Fragments.
        val density = resources.displayMetrics.density
        val scrollView = ScrollView(requireContext()).apply {
            isFillViewport = true
            scrollBarStyle = View.SCROLLBARS_INSIDE_OVERLAY
            scrollBarSize = (3 * density).toInt()
            isScrollbarFadingEnabled = true
            scrollBarDefaultDelayBeforeFade = 400
            scrollBarFadeDuration = 250
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
                setVerticalScrollbarThumbDrawable(GradientDrawable().apply {
                    shape = GradientDrawable.RECTANGLE
                    setColor(android.graphics.Color.argb(100, 26, 159, 255))
                    cornerRadius = 4 * density
                })
            }
            addView(
                composeView,
                ViewGroup.LayoutParams(
                    ViewGroup.LayoutParams.MATCH_PARENT,
                    ViewGroup.LayoutParams.WRAP_CONTENT,
                ),
            )
        }
        // Outer FrameLayout with right margin so the scrollbar never touches the screen edge.
        // Background matches the content area so the gap is invisible.
        return FrameLayout(requireContext()).apply {
            setBackgroundColor(android.graphics.Color.parseColor("#0F0F12"))
            addView(
                scrollView,
                FrameLayout.LayoutParams(
                    FrameLayout.LayoutParams.MATCH_PARENT,
                    FrameLayout.LayoutParams.MATCH_PARENT,
                ).apply { marginEnd = (10 * density).toInt() },
            )
        }
    }

    override fun onResume() {
        super.onResume()
        refresh()
    }

    // Helpers
    private fun refresh() {
        val ctx = context ?: return
        storeState = StoreState(
            isSteamLoggedIn = SteamService.isLoggedIn,
            isEpicLoggedIn  = EpicAuthManager.isLoggedIn(ctx),
            isGogLoggedIn   = GOGAuthManager.isLoggedIn(ctx),
            sharedFolder    = PrefManager.useSingleDownloadFolder,
            downloadSpeed   = PrefManager.downloadSpeed,
            downloadServer           = PrefManager.cellId,
            downloadServerManuallySet = PrefManager.cellIdManuallySet,
            defaultFolder   = resolveUri(PrefManager.defaultDownloadFolder, ctx),
            steamFolder     = resolveUri(PrefManager.steamDownloadFolder,   ctx),
            epicFolder      = resolveUri(PrefManager.epicDownloadFolder,    ctx),
            gogFolder       = resolveUri(PrefManager.gogDownloadFolder,     ctx),
            amazonFolder    = resolveUri(PrefManager.amazonDownloadFolder,  ctx),
        )
    }

    private fun loadServerOptions(): List<Pair<Int, String>> {
        return try {
            val json = requireContext().assets.open("steam_regions.json")
                .bufferedReader().use { it.readText() }
            val obj = JSONObject(json)
            obj.keys().asSequence()
                .mapNotNull { key -> key.toIntOrNull()?.let { id -> id to obj.getString(key) } }
                .sortedBy { it.first }
                .toList()
        } catch (_: Exception) {
            listOf(0 to "Automatic")
        }
    }

    private fun persistUri(uri: Uri) {
        try {
            requireContext().contentResolver.takePersistableUriPermission(
                uri,
                Intent.FLAG_GRANT_READ_URI_PERMISSION or Intent.FLAG_GRANT_WRITE_URI_PERMISSION,
            )
        } catch (_: SecurityException) {}
    }

    private fun resolveUri(uriStr: String, ctx: android.content.Context): String {
        if (uriStr.isEmpty()) return ""
        return try {
            FileUtils.getFilePathFromUri(ctx, Uri.parse(uriStr)) ?: uriStr
        } catch (_: Exception) { uriStr }
    }
}
